metaflow/includefile.py

```diff
@@ -8,259 +9,209 @@
-    #      of the persisted file
-    #  - The artifact will be persisted prior to any run (for non-scheduled runs through persist_constants)
-    #    + This will therefore persist a simple string
-    #  - When the parameter is loaded again, the load_parameter in the IncludeFile class will get called
-    #    which will download and return the bytes of the persisted file.
+
     def __init__(self, is_text, encoding):
         self._is_text = is_text
         self._encoding = encoding
 
     def convert(self, value, param, ctx):
-        if callable(value):
-            # Already a correct type
+        # Click can call convert multiple times so we need to make sure to only
+        # convert once. This function will return a DelayedEvaluationParameter
+        # (if it needs to still perform an upload) or an IncludedFile if not
+        if isinstance(value, (DelayedEvaluationParameter, IncludedFile)):
             return value
 
-        value = os.path.expanduser(value)
-        ok, file_type, err = LocalFile.is_file_handled(value)
-        if not ok:
-            self.fail(err)
-        if file_type is None:
-            # Here, we need to store the file
-            return lambda is_text=self._is_text, encoding=self._encoding, value=value, ctx=parameters.context_proto: LocalFile(
-                is_text, encoding, value
-            )(
-                ctx
+        # Value will be a string containing one of two things:
+        #  - a JSON blob indicating that the file has already been uploaded. The
+        #    scenario this happens in is as follows:
+        #      + `step-functions create` is called and the IncludeFile has a default
+        #        value. At the time of creation, the file is uploaded and a URL is
+        #        returned; this URL is packaged in a blob by Uploader and passed to
+        #        step-functions as the value of the parameter.
+        #      + when the step function actually runs, the value is passed to click
+        #        through METAFLOW_INIT_XXX; this value is the one returned above
+        #    This will be scenario A.
+        #  - A path. The path can either be:
+        #      + <prefix>://<something> like s3://foo/bar or local:///foo/bar
+        #      + an actual path to a local file like /foo/bar
+        #    In the first case, we just store an *external* reference to it (so we
+        #    won't upload anything). In the second case, we will want to upload something
+        #    but we only do that in the DelayedEvaluationParameter step.
+        #    This will be scenario B subdivided in B.1 for the <prefix> case and B.2
+        #    for the other case.
+
+        # This should not happen but being safe
+        if value is None:
+            return None
+
+        # ctx can be one of two things:
+        #  - the click context (when called normally)
+        #  - the ParameterContext (when called through _eval_default)
```

Comment:
It's hard to keep track of function names in code comments - somebody renames `_eval_default` etc.